packages/components/src/query-controls/types.ts

```diff
@@ -0,0 +1,131 @@
+/**
+ * External dependencies
+ */
+import type { Key } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import type { FormTokenFieldProps } from '../form-token-field/types';
+import type { TreeSelectProps } from '../tree-select/types';
+
+export type Entity = {
+	id: number;
+	name: string;
+	parent?: number | null;
+	value?: string;
+};
+
+export type EntityForTree = Omit< Entity, 'id' > & {
+	id: string;
+};
+
+export type TermsWithChildren = Array<
+	EntityForTree & { children: EntityForTree[] }
+>;
+
+export type Category = {
```

Comment:
Sure, how does https://github.com/WordPress/gutenberg/pull/46721/commits/30748b9efdc9c8d47c12b68ac035d75454e4fa54 look?

Reply:
So, I took a deeper look.

I had a look at the WordPress API reference — authors should be receiving [users data](https://developer.wordpress.org/rest-api/reference/users/), while categories should be receiving [categories data](https://developer.wordpress.org/rest-api/reference/categories/).

In that sense, I actually think it's ok for this component to define different types for `Author` and `Category` as you initially proposed in this PR (sorry for the turnaround!). Also, since we don't want the components library to be tightly coupled to the WordPress backend, it's ok to keep those types as lean as possible (as opposed to listing all props from those entities).

These APIs also confirm that `id` is supposed to be a `number`.

I also don't think that `value` really belongs to the `Category` type, and therefore I think that for now we should just `@ts-ignore` its usage in `index.tsx` to keep types clean.

I tweaked the data in the Storybook examples to adhere to the new type specs.

Finally, I also slightly reworked the `EntityForTree` type, and replaced it with a `TermWithParentAndChildren` type.

<details>

<summary>Here's a diff with these changes applied, what do you think?</summary>

```diff
diff --git a/packages/components/src/query-controls/index.tsx b/packages/components/src/query-controls/index.tsx
index f940540963..91769410b4 100644
--- a/packages/components/src/query-controls/index.tsx
+++ b/packages/components/src/query-controls/index.tsx
@@ -148,6 +148,10 @@ export function QueryControls( {
 								props.selectedCategories &&
 								props.selectedCategories.map( ( item ) => ( {
 									id: item.id,
+									// Keeping the fallback to `item.value` for legacy reasons,
+									// even if items of `selectedCategories` should not have a
+									// `value` property.
+									// @ts-expect-error
 									value: item.name || item.value,
 								} ) )
 							}
diff --git a/packages/components/src/query-controls/stories/index.tsx b/packages/components/src/query-controls/stories/index.tsx
index 54354ed537..1bb7f26cdf 100644
--- a/packages/components/src/query-controls/stories/index.tsx
+++ b/packages/components/src/query-controls/stories/index.tsx
@@ -13,7 +13,7 @@ import { useState } from '@wordpress/element';
  */
 import QueryControls from '..';
 import type {
-	Entity,
+	Category,
 	QueryControlsWithSingleCategorySelectionProps,
 	QueryControlsWithMultipleCategorySelectionProps,
 } from '../types';
@@ -76,7 +76,7 @@ export const Default: ComponentStory< typeof QueryControls > = ( args ) => {
 						? props.categorySuggestions?.[ token ]
 						: token;
 				} )
-				.filter( Boolean ) as Array< Required< Entity > >;
+				.filter( Boolean ) as Array< Required< Category > >;
 
 			setOwnSelectedCategories( allCategories );
 		};
@@ -125,20 +125,19 @@ Default.args = {
 		TypeScript: {
 			id: 11,
 			name: 'TypeScript',
+			parent: 0,
 		},
 		JavaScript: {
 			id: 12,
 			name: 'JavaScript',
+			parent: 0,
 		},
 	},
 	selectedCategories: [
 		{
 			id: 11,
-			value: 'TypeScript',
-		},
-		{
-			id: 12,
-			value: 'JavaScript',
+			name: 'TypeScript',
+			parent: 0,
 		},
 	],
 	numberOfItems: 5,
@@ -194,10 +193,12 @@ SelectSingleCategory.args = {
 		{
 			id: 11,
 			name: 'TypeScript',
+			parent: 0,
 		},
 		{
 			id: 12,
 			name: 'JavaScript',
+			parent: 0,
 		},
 	],
 	selectedCategoryId: 11,
diff --git a/packages/components/src/query-controls/terms.ts b/packages/components/src/query-controls/terms.ts
index aed7c699cc..3802e05d84 100644
--- a/packages/components/src/query-controls/terms.ts
+++ b/packages/components/src/query-controls/terms.ts
@@ -6,7 +6,12 @@ import { groupBy } from 'lodash';
 /**
  * Internal dependencies
  */
-import type { Entity, EntityForTree, TermsWithChildren } from './types';
+import type {
+	Author,
+	Category,
+	TermWithParentAndChildren,
+	TermsByParent,
+} from './types';
 
 /**
  * Returns terms in a tree form.
@@ -15,21 +20,27 @@ import type { Entity, EntityForTree, TermsWithChildren } from './types';
  *
  * @return Terms in tree format.
  */
-export function buildTermsTree( flatTerms: readonly Entity[] ) {
-	const flatTermsWithParentAndChildren = flatTerms.map( ( term ) => {
-		return {
-			children: [],
-			parent: null,
-			...term,
-			id: String( term.id ),
-		};
-	} );
+export function buildTermsTree( flatTerms: readonly ( Author | Category )[] ) {
+	const flatTermsWithParentAndChildren: TermWithParentAndChildren[] =
+		flatTerms.map( ( term ) => {
+			return {
+				children: [],
+				parent: null,
+				...term,
+				id: String( term.id ),
+			};
+		} );
 
-	const termsByParent = groupBy( flatTermsWithParentAndChildren, 'parent' );
+	const termsByParent: TermsByParent = groupBy(
+		flatTermsWithParentAndChildren,
+		'parent'
+	);
 	if ( termsByParent.null && termsByParent.null.length ) {
 		return flatTermsWithParentAndChildren;
 	}
-	const fillWithChildren = ( terms: EntityForTree[] ): TermsWithChildren => {
+	const fillWithChildren = (
+		terms: TermWithParentAndChildren[]
+	): TermWithParentAndChildren[] => {
 		return terms.map( ( term ) => {
 			const children = termsByParent[ term.id ];
 			return {
diff --git a/packages/components/src/query-controls/types.ts b/packages/components/src/query-controls/types.ts
index fe5d1a70af..6086d35e2a 100644
--- a/packages/components/src/query-controls/types.ts
+++ b/packages/components/src/query-controls/types.ts
@@ -4,37 +4,42 @@
 import type { FormTokenFieldProps } from '../form-token-field/types';
 import type { TreeSelectProps } from '../tree-select/types';
 
-export type Entity = {
+export type Author = {
 	id: number;
 	name: string;
-	parent?: number | null;
-	value?: string;
 };
 
-export type EntityForTree = Omit< Entity, 'id' > & {
+export type Category = {
+	id: number;
+	name: string;
+	parent: number;
+};
+
+export type TermWithParentAndChildren = {
 	id: string;
+	name: string;
+	parent: number | null;
+	children: TermWithParentAndChildren[];
 };
 
-export type TermsWithChildren = Array<
-	EntityForTree & { children: EntityForTree[] }
->;
+export type TermsByParent = Record< string, TermWithParentAndChildren[] >;
 
 export type CategorySelectProps = Pick<
 	TreeSelectProps,
 	'label' | 'noOptionLabel'
 > & {
-	categoriesList: Entity[];
+	categoriesList: Category[];
 	onChange: ( newCategory: string ) => void;
-	selectedCategoryId?: Entity[ 'id' ];
+	selectedCategoryId?: Category[ 'id' ];
 };
 
 export type AuthorSelectProps = Pick<
 	TreeSelectProps,
 	'label' | 'noOptionLabel'
 > & {
-	authorList?: Entity[];
+	authorList?: Author[];
 	onChange: ( newAuthor: string ) => void;
-	selectedAuthorId?: Entity[ 'id' ];
+	selectedAuthorId?: Author[ 'id' ];
 };
 
 type Order = 'asc' | 'desc';
@@ -123,16 +128,13 @@ export type QueryControlsWithMultipleCategorySelectionProps =
 		 * An array of category names, renders a `FormTokenField` component
 		 * when passed in conjunction with `onCategoryChange`.
 		 */
-		categorySuggestions?: { [ categoryName: Entity[ 'name' ] ]: Entity };
+		categorySuggestions?: {
+			[ categoryName: Category[ 'name' ] ]: Category;
+		};
 		/**
 		 * The selected categories for the `categorySuggestions`.
 		 */
-		selectedCategories?: Array<
-			Omit< Entity, 'name' > & {
-				name?: string;
-				value: string;
-			}
-		>;
+		selectedCategories?: Category[];
 		/**
 		 * A function that receives the new category value.
 		 * If this is not specified, the category controls are not included.
```

</details>

Reply:
Thanks, committed your diff verbatim in https://github.com/WordPress/gutenberg/pull/46721/commits/a08a2404a305e85115e3e8f5916e379c696d598e.

Good idea to keep `item.value` for legacy reasons, but exclude it from the types.